common/logger/derive/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+            Some(ValueType::Display) => &from_display,
+            Some(ValueType::Debug) => &from_debug,
+            Some(ValueType::Serde) | None => &from_serde,
+        };
+        if f.inner_ty.is_some() {
+            quote! {
+                if let Some(#ident) = &self.#ident {
+                    #visitor.visit_pair(#key_new(#ident_str), #from_fn(#ident));
+                }
+            }
+        } else {
+            quote! {
+                #visitor.visit_pair(#key_new(#ident_str), #from_fn(&self.#ident));
+            }
+        }
+    });
+
+    // Build the output, possibly using quasi-quotation
+    let expanded = quote! {
+        impl #impl_generics #name #ty_generics #where_clause {
+            #(#setters)*
+
+            pub fn into_struct_log(self) -> ::libra_logger::StructuredLogEntry {
+                ::libra_logger::StructuredLogEntry::new_unnamed().schema(self)
+            }
+        }
+
+        impl #impl_generics ::libra_logger::Schema for #name #ty_generics #where_clause {
+            fn visit(&self, visitor: &mut dyn ::libra_logger::Visitor) {
+                #(#visits)*
+            }
+        }
+    };
+
+    // Hand the output tokens back to the compiler
+    TokenStream::from(expanded)
+}
+
+#[derive(Debug)]
+enum ValueType {
+    Debug,
+    Display,
+    Serde,
+}
+
+#[derive(Debug)]
+struct StructField {
+    value_type: Option<ValueType>,
+    ident: Option<Ident>,
+    ty: Type,
+    /// Indicates that the type is wrapped by an Option type
+    inner_ty: Option<Type>,
+}
+
+fn extract_internal_type(ty: &Type) -> Option<&Type> {
+    if let Type::Path(TypePath {
+        qself: None,
+        path: Path { segments, .. },
+    }) = ty
+    {
+        if let Some(PathSegment { ident, arguments }) = segments.first() {
+            // Extract the inner type if it is "Option"
+            if ident == "Option" {
+                if let PathArguments::AngleBracketed(AngleBracketedGenericArguments {
+                    args, ..
+                }) = arguments
+                {
+                    if let Some(GenericArgument::Type(ty)) = args.first() {
+                        return Some(ty);
+                    }
+                }
+            }
+        }
+    }
+
+    None
+}
+
+fn extract_attr(attrs: &[Attribute]) -> Option<ValueType> {
+    for attr in attrs {
+        if let Meta::List(MetaList { path, nested, .. }) = attr.parse_meta().unwrap() {
+            for segment in path.segments {
+                // Only handle schema attrs
+                if segment.ident == "schema" {
+                    for meta in &nested {
+                        let path = if let NestedMeta::Meta(Meta::Path(path)) = meta {
+                            path
+                        } else {
+                            panic!("unsupported schema attribute");
+                        };
+
+                        match path.segments.first().unwrap().ident.to_string().as_ref() {
```

Comment:
Maybe we want to case insensitive the match?

Reply:
I don't know if that's something that most proc-macros do, we can take a look at serde and see but i'm not overly concerned